### ISSUE TITLE
Count tokens used in retried requests due to Pydantic validation failures

### ISSUE THREAD

**Issue by Elijas** [instructor-ai/instructor#715]
**Is your feature request related to a problem? Please describe.**
I'm frustrated that only the last request tokens are included when retrying to extract the Pydantic model after a validation failure. This limits the effectiveness of retries, especially when using multiple LLMs like Litellm and Anthropic. It would be helpful to have a complete token count for all retried requests to better manage context and improve accuracy.

**Describe the solution you'd like**
I would like the retry logic to count and include tokens from all previous requests, not just the last one. This will ensure a comprehensive context for retries, improving the effectiveness of the model extraction process and better handling validation failures.

**Describe alternatives you've considered**
- Using the max_retries Parameter: This currently only includes the last request tokens, which doesn't provide a full context.
- Custom Retry Logic with Tenacity: This approach could work but requires manually integrating token counts from all previous requests, which is complex and not straightforward without built-in support.

**Is your feature request related to a problem? Please describe.**
I'm frustrated that only the last request tokens are included when retrying to extract the Pydantic model after a validation failure. This limits the effectiveness of retries, especially when using multiple LLMs like Litellm and Anthropic. It would be helpful to have a complete token count for all retried requests to better manage context and improve accuracy.

**Additional context**
Including the token count from all retried requests will help maintain full context, making the retry process more reliable and accurate. This enhancement would be particularly beneficial for users leveraging multiple LLMs, ensuring comprehensive error handling and model extraction.

Thank you for considering this feature request. It would greatly improve the utility and reliability of the Instructor library.

---

I'm using `Instructor -> LiteLLM (library, not proxy) -> Anthropic Claude` stack

**Comment by jxnl** [comment]
Good call. happy to take a PR; otherwise, most big use cases I've seen with the instructor already lean on third-party systems to track the token usage.

**Comment by Elijas** [comment]
> Good call. happy to take a PR; otherwise, most big use cases I've seen with the instructor already lean on third-party systems to track the token usage.

third party systems:
litellm proxy? Or youve seen something else

**Comment by jxnl** [comment]
helicone, langsmith, logfire, parea,

**Comment by ionflow** [comment]
I need this as well. Will look at implementing something.

**Comment by cpfiffer** [comment]
+1 on this one. It seems like I'd need to write custom hooks to track each retry?

**Comment by cpfiffer** [comment]
For anyone who stumbles on this, I wrote a hook to track this:

```python
from rich import print

client.clear()

class UsageTracker:
    input_tokens: list[int] = []
    output_tokens: list[int] = []
    total_tokens: list[int] = []

    def track(self, usage):
        self.input_tokens.append(usage.prompt_tokens)
        self.output_tokens.append(usage.completion_tokens)
        self.total_tokens.append(usage.total_tokens)

    def clear(self):
        self.input_tokens.clear()
        self.output_tokens.clear()
        self.total_tokens.clear()

tracker = UsageTracker()

def log_completion_kwargs(*args, **kwargs):
    usage = args[0].usage
    tracker.track(usage)

client.on("completion:response", log_completion_kwargs)

# Normal instructor calls
# . . .
# . . .

print(tracker.input_tokens)
print(tracker.output_tokens)

```

**Comment by jxnl** [comment]
Closing as part of repository maintenance for issues created before 2025.
